sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala

```diff
@@ -152,26 +153,54 @@ class GpuSingleDirectoryDataWriter(
   override def write(batch: ColumnarBatch): Unit = {
     var needToCloseBatch = true
     try {
-      // TODO: This does not handle batch splitting to make sure maxRecordsPerFile is not exceeded.
-      if (description.maxRecordsPerFile > 0 && recordsInFile >= description.maxRecordsPerFile) {
-        fileCounter += 1
-        assert(fileCounter < MAX_FILE_COUNTER,
-          s"File counter $fileCounter is beyond max value $MAX_FILE_COUNTER")
+      withResource(GpuColumnVector.from(batch)) {table => 
+        val numRows = table.getRowCount()
+        val maxRecordsPerFile = description.maxRecordsPerFile.toInt
+        val fillOldFile = recordsInFile < maxRecordsPerFile
+        val splitIndexes = (maxRecordsPerFile > 0, fillOldFile) match {
+          case (false, _) => IndexedSeq.empty
+          case (true, false) =>
+            (1 until ((numRows + maxRecordsPerFile - 1) / maxRecordsPerFile).toInt)
+              .map(i => i * maxRecordsPerFile)
+          case (true, true) => {
+            val filledUp = maxRecordsPerFile - recordsInFile.toInt
+            val remain = numRows - filledUp
+            (0 until ((remain + maxRecordsPerFile - 1) / maxRecordsPerFile).toInt)
+              .map(i => filledUp + i * maxRecordsPerFile)
+          }
+        }
 
-        newOutputWriter()
+        val dataTypes = (0 until batch.numCols()).map(i => batch.column(i).dataType()).toArray
+
+        var needNewWriter = !fillOldFile
+
+        withResource(table.contiguousSplit(splitIndexes: _*)) {tabs =>
```

Comment:
We don't want to pay for an unnecessary contiguous split when we don't have a max record per file limit (which is the common case).

Reply:
When we don't set the `maxRecordPerFile`, the `splitIndexes` will be empty:
```scala
        val splitIndexes = (maxRecordsPerFile > 0, fillOldFile) match {
          case (false, _) => IndexedSeq.empty
...
```

Reply:
I agree that we do not need to split if `maxRecordsPerFile <=0`, we can save some time to avoid this JNI invoke: `table.contiguousSplit`

Reply:
Done.